OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs

```diff
@@ -44,220 +51,258 @@ public class MapPreviewLogic : ChromeLogic
-			preview.OnMouseDown = mi => onMouseDown(preview, getMap().Map, mi);
-			preview.SpawnOccupants = getSpawnOccupants;
-			preview.DisabledSpawnPoints = getDisabledSpawnPoints;
-			preview.ShowUnoccupiedSpawnpoints = showUnoccupiedSpawnpoints;
-
-			var titleLabel = parent.GetOrNull<LabelWithTooltipWidget>("MAP_TITLE");
-			if (titleLabel != null)
+			var (map, serverStatus) = getMap();
+			PreviewStatus? status = null;
+
+			if (map == MapCache.UnknownMap)
+				status = PreviewStatus.Unknown;
+			else
 			{
-				titleLabel.IsVisible = () => getMap().Map != MapCache.UnknownMap;
-				var font = Game.Renderer.Fonts[titleLabel.Font];
-				var title = new CachedTransform<MapPreview, string>(m =>
+				switch (map.Status)
 				{
-					var truncated = WidgetUtils.TruncateText(m.Title, titleLabel.Bounds.Width, font);
-
-					if (m.Title != truncated)
-						titleLabel.GetTooltipText = () => m.Title;
-					else
-						titleLabel.GetTooltipText = null;
-
-					return truncated;
-				});
-				titleLabel.GetText = () => title.Update(getMap().Map);
+					case MapStatus.Available:
+						if (serverStatus.HasFlag(Session.MapStatus.Playable))
+							status = PreviewStatus.Playable;
+						else if (serverStatus.HasFlag(Session.MapStatus.Incompatible))
+							status = PreviewStatus.Incompatible;
+						else if (serverStatus.HasFlag(Session.MapStatus.Validating))
+							status = PreviewStatus.Validating;
+						else
+							return;
+						break;
+					case MapStatus.DownloadAvailable:
+						if (mapUpdateAvailable)
+							status = PreviewStatus.UpdateDownloadAvailable;
+						else
+							status = PreviewStatus.DownloadAvailable;
+						break;
+					case MapStatus.Searching:
+						status = PreviewStatus.Searching;
+						break;
+					case MapStatus.Unavailable:
+						if (mapUpdateAvailable)
+							status = PreviewStatus.UpdateAvailable;
+						else
+							status = PreviewStatus.Unavailable;
+						break;
+					case MapStatus.DownloadError:
+						status = PreviewStatus.DownloadError;
+						break;
+					case MapStatus.Downloading:
+						status = PreviewStatus.Downloading;
+						break;
+				}
 			}
 
-			var typeLabel = parent.GetOrNull<LabelWidget>("MAP_TYPE");
-			if (typeLabel != null)
+			if (status != null && status != currentStatus)
 			{
-				var type = new CachedTransform<MapPreview, string>(m => m.Categories.FirstOrDefault() ?? "");
-				typeLabel.GetText = () => type.Update(getMap().Map);
-			}
+				if (status == PreviewStatus.Unavailable)
+				{
+					triggered = true;
+					retryButton.OnClick();
+					if (mapUpdateAvailable)
+						status = PreviewStatus.UpdateAvailable;
+				}
+				else if (status == PreviewStatus.UpdateAvailable)
+				{
+					triggered = true;
+					retryButton.OnClick();
+					if (!mapUpdateAvailable)
+						status = PreviewStatus.Unavailable;
+				}
+				else if (status == PreviewStatus.Searching && triggered)
+				{
+					// We don't switch to Searching status as we don't want to repeatedly trigger retryButton
```

Comment:
I don't understand this comment. We are already in the searching status?

Reply:
We aren't, the status setter is below and we are `return`ing before it is triggered

Reply:
This is very confusing, the if check above literally says `status == PreviewStatus.Searching`.

Reply:
I agree that it is confusing, that's why I added a comment. I guess I could make it a bit more detailed